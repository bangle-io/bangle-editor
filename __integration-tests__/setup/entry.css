@import '@banglejs/core/style.css';
