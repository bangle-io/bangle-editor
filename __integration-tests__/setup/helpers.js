const prettier = require('prettier');
const os = require('os');

const pmRoot = '#pm-root';
const ctrlKey = os.platform() === 'darwin' ? 'Meta' : 'Control';

module.exports = {
  mountEditor,
  getEditorState,
  pressRight,
  pressLeft,
  debug,
  ctrlKey,
  pmRoot,
  getDoc,
  sleep,
};

function frmt(doc) {
  return prettier.format(doc.toString(), {
    semi: false,
    parser: 'babel',
    printWidth: 36,
    singleQuote: true,
  });
}

async function mountEditor(page, props) {
  await page.waitForSelector(pmRoot, { timeout: 500 });
  await page.waitForSelector('.ProseMirror', { timeout: 500 });
}

async function getEditorState(page) {
  return page.evaluate(() => {
    return window.editor.view.state.toJSON();
  });
}

async function pressRight() {
  await page.keyboard.press('ArrowRight');
}

async function pressLeft() {
  await page.keyboard.press('ArrowLeft');
}

function debug() {
  return jestPuppeteer.debug();
}

function sleep(t = 20) {
  return new Promise((res) => setTimeout(res, t));
}

async function getDoc(page) {
  return page
    .evaluate(() => {
      return window.editor.view.state.doc.toString();
    })
    .then(frmt);
}
