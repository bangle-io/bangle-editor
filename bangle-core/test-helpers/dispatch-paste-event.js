import browser from 'bangle-core/utils/browser';
import { createEvent } from './create-event';

/**
 *
 *
 * Dispatch a paste event on the given ProseMirror instance
 *
 * Usage:
 *     dispatchPasteEvent(pm, {
 *         plain: 'copied text'
 *     });
 */
export function dispatchPasteEvent(editorView, content) {
  const event = createEvent('paste');
  const clipboardData = {
    getData(type) {
      if (type === 'text/plain') {
        return content.plain;
      }
      if (type === 'text/html') {
        return content.html;
      }
      if (type === 'text/uri-list') {
        return content['uri-list'];
      }
      return;
    },
    types: content.types || [],
    files: content.files || [],
    items: content.items || null,
  };
  // Reason: https://github.com/ProseMirror/prosemirror-view/blob/9d2295d03c2d17357213371e4d083f0213441a7e/bangle-play/input.js#L379-L384
  if ((browser.ie && browser.ie_version < 15) || browser.ios) {
    return false;
  }
  Object.defineProperty(event, 'clipboardData', { value: clipboardData });

  editorView.dispatchEvent(event);
  return event;
}
