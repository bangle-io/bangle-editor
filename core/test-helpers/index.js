import { NodeSelection } from 'prosemirror-state';

export * from './commands-helpers';
export * from './jest-helpers';
export * from './keyboard';
export * from './render-test-editor';
export * from './schema-builders';
export * from './dispatch-paste-event';
export * from './create-event';
export * from './selection-helpers';

export const selectNodeAt = (view, pos) => {
  const tr = view.state.tr;
  view.dispatch(tr.setSelection(NodeSelection.create(tr.doc, pos)));
};
