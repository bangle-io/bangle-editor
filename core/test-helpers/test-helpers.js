import { NodeSelection } from 'prosemirror-state';
import { Fragment, Slice } from 'prosemirror-model';

export * from './commands-helpers';
export * from './jest-helpers';
export * from './keyboard';
export * from './render-test-editor';
export * from './schema-builders';
export * from './dispatch-paste-event';
export * from './create-event';
export * from './selection-helpers';

export const selectNodeAt = (view, pos) => {
  const tr = view.state.tr;
  view.dispatch(tr.setSelection(NodeSelection.create(tr.doc, pos)));
};

/**
 *
 * @param {*} schema
 * @param {*} psxArray An array of psx nodes eg. [<para>hi</para>, <para>bye</para>]
 */
export const createPSXFragment = (schema, psxArray) => {
  return Fragment.fromArray(psxArray.map((r) => r(schema)));
};
