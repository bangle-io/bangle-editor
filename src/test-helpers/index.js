export * from './commands-helpers';
export * from './jest-helpers';
export * from './keyboard';
export * from './render-test-editor';
export * from './test-builders';
