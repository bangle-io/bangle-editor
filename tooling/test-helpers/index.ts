import { EditorView, Fragment, NodeSelection, Schema } from '@bangle.dev/pm';

export * from './create-event';
export * from './dispatch-paste-event';
export * from './jest-helpers';
export * from './keyboard';
export * from './render-test-editor';
export * from './schema-builders';
export * from './selection-helpers';

export const selectNodeAt = (view: EditorView, pos: number) => {
  const tr = view.state.tr;
  view.dispatch(tr.setSelection(NodeSelection.create(tr.doc, pos)));
};

/**
 *
 * @param {*} schema
 * @param {*} psxArray An array of psx nodes eg. [<para>hi</para>, <para>bye</para>]
 */
export const createPSXFragment = (schema: Schema, psxArray: any[]) => {
  return Fragment.fromArray(psxArray.map((r) => r(schema)));
};
