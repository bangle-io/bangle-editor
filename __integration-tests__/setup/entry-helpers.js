import React from 'react';
import reactDOM from 'react-dom';
import { ReactEditorView, useEditorState } from '@banglejs/react';
import { SpecRegistry } from '@banglejs/core/spec-registry';
import {
  defaultPlugins,
  defaultSpecs,
} from '@banglejs/core/test-helpers/default-components';

export function setupReactEditor({
  specRegistry,
  plugins = defaultPlugins,
  renderNodeViews,
  id = 'pm-root',
} = {}) {
  const element = document.createElement('div');
  window.document.body.appendChild(element);
  element.setAttribute('id', 'root');
  if (!(specRegistry instanceof SpecRegistry)) {
    specRegistry = new SpecRegistry(defaultSpecs(specRegistry));
  }
  reactDOM.render(
    <App opts={{ specRegistry, plugins, renderNodeViews, id }} />,
    element,
  );
}

function App({ opts: { specRegistry, plugins, renderNodeViews, id } }) {
  const onEditorReady = (_editor) => {
    window.editor = _editor;
  };

  window.dispatcher = (command) => {
    return command(
      window.editor.view.state,
      window.editor.view.dispatch,
      window.editor.view,
    );
  };

  const editorState = useEditorState({ specRegistry, plugins });

  return (
    <ReactEditorView
      id={id}
      editorState={editorState}
      onReady={onEditorReady}
      renderNodeViews={renderNodeViews}
    />
  );
}
