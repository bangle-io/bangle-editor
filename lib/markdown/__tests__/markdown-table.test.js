/**
 * @jest-environment jsdom
 */

/** @jsx psx */
import fs from 'fs/promises';
import path from 'path';

import { psx } from '@bangle.dev/test-helpers';

import { parse, serialize } from './setup';

describe('Markdown table parsing', () => {
  test('table1.md: a simple table', async () => {
    const doc = (
      <doc>
        <heading>My Table</heading>
        <table>
          <table_row>
            <table_header align="left">
              <para>NOTE:</para>
            </table_header>
          </table_row>
          <table_row>
            <table_cell align="left">
              <para>Work in progress</para>
            </table_cell>
          </table_row>
        </table>
      </doc>
    );

    const md = await fs.readFile(
      path.join(__dirname, './fixtures/table1.md'),
      'utf-8',
    );
    const resultDoc = await parse(md);
    expect(resultDoc).toEqualDocument(doc);

    // check serializing
    expect(await serialize(doc)).toMatchSnapshot();
    const parsedDoc = await parse(await serialize(doc));
    expect(parsedDoc).toEqualDocument(doc);
  });

  test('table2.md: two simple table with two columns', async () => {
    // there two tables that differ with their heading syntax
    const t1 = (
      <table>
        <table_row>
          <table_header>
            <para>Header 1</para>
          </table_header>
          <table_header>
            <para>Header 2</para>
          </table_header>
        </table_row>
        <table_row>
          <table_cell>
            <para>hello</para>
          </table_cell>
          <table_cell>
            <para>world</para>
          </table_cell>
        </table_row>
      </table>
    );
    const doc = (
      <doc>
        <heading>Table 2</heading>
        {t1}
        {t1}
      </doc>
    );

    const md = await fs.readFile(
      path.join(__dirname, './fixtures/table2.md'),
      'utf-8',
    );
    const resultDoc = await parse(md);

    expect(resultDoc).toEqualDocument(doc);

    // check serializing
    expect(await serialize(doc)).toMatchSnapshot();
    const parsedDoc = await parse(await serialize(doc));
    expect(parsedDoc).toEqualDocument(doc);
  });

  test('table3.md: with alignment', async () => {
    const doc = (
      <doc>
        <heading>Table 3</heading>
        <table>
          <table_row>
            <table_header>
              <para>Default</para>
            </table_header>
            <table_header align="left">
              <para>Left</para>
            </table_header>
            <table_header align="center">
              <para>Center</para>
            </table_header>
            <table_header align="right">
              <para>Right</para>
            </table_header>
          </table_row>
          <table_row>
            <table_cell>
              <para>x</para>
            </table_cell>
            <table_cell align="left">
              <para>x</para>
            </table_cell>
            <table_cell align="center">
              <para>x</para>
            </table_cell>
            <table_cell align="right">
              <para>x</para>
            </table_cell>
          </table_row>
        </table>
      </doc>
    );

    const md = await fs.readFile(
      path.join(__dirname, './fixtures/table3.md'),
      'utf-8',
    );
    const resultDoc = await parse(md);

    expect(resultDoc).toEqualDocument(doc);

    // check serializing
    expect(await serialize(doc)).toMatchSnapshot();
    const parsedDoc = await parse(await serialize(doc));
    expect(parsedDoc).toEqualDocument(doc);
  });

  test('table4.md', async () => {
    const doc = (
      <doc>
        <heading>Table 4</heading>
        <table>
          <table_row>
            <table_header align="center">
              <para>foo</para>
            </table_header>
            <table_header align="right">
              <para>goo</para>
            </table_header>
          </table_row>
          <table_row>
            <table_cell align="center">
              <para>bar</para>
            </table_cell>
            <table_cell align="right">
              <para>baz</para>
            </table_cell>
          </table_row>
        </table>
      </doc>
    );

    const md = await fs.readFile(
      path.join(__dirname, './fixtures/table4.md'),
      'utf-8',
    );
    const resultDoc = await parse(md);
    expect(resultDoc).toEqualDocument(doc);

    // check serializing
    expect(await serialize(doc)).toMatchSnapshot();
    const parsedDoc = await parse(await serialize(doc));
    expect(parsedDoc).toEqualDocument(doc);
  });

  test('table5.md', async () => {
    const doc = (
      <doc>
        <heading>Table 5</heading>
        <table>
          <table_row>
            <table_header>
              <para>f|oo</para>
            </table_header>
          </table_row>
          <table_row>
            <table_cell>
              <para>
                b <code>\|</code> az
              </para>
            </table_cell>
          </table_row>
          <table_row>
            <table_cell>
              <para>
                b <bold>|</bold> im
              </para>
            </table_cell>
          </table_row>
        </table>
      </doc>
    );

    const md = await fs.readFile(
      path.join(__dirname, './fixtures/table5.md'),
      'utf-8',
    );
    const resultDoc = await parse(md);
    expect(resultDoc).toEqualDocument(doc);

    // check serializing
    expect(await serialize(doc)).toMatchSnapshot();
    const parsedDoc = await parse(await serialize(doc));
    expect(parsedDoc).toEqualDocument(doc);
  });

  test('table6.md', async () => {
    const doc = (
      <doc>
        <heading>Table 6</heading>
        <table>
          <table_row>
            <table_header>
              <para>abc</para>
            </table_header>
            <table_header>
              <para>def</para>
            </table_header>
          </table_row>
          <table_row>
            <table_cell>
              <para>bar</para>
            </table_cell>
            <table_cell>
              <para>baz</para>
            </table_cell>
          </table_row>
        </table>
        <blockquote>
          <para>bar</para>
        </blockquote>
      </doc>
    );

    const md = await fs.readFile(
      path.join(__dirname, './fixtures/table6.md'),
      'utf-8',
    );
    const resultDoc = await parse(md);
    expect(resultDoc).toEqualDocument(doc);

    // check serializing
    expect(await serialize(doc)).toMatchSnapshot();
    const parsedDoc = await parse(await serialize(doc));
    expect(parsedDoc).toEqualDocument(doc);
  });

  test('table7 missing column cell', async () => {
    const doc = (
      <doc>
        <heading>Table 7</heading>
        <para>| abc | def | | --- | | bar |</para>
      </doc>
    );

    const md = await fs.readFile(
      path.join(__dirname, './fixtures/table7.md'),
      'utf-8',
    );
    const resultDoc = await parse(md);
    expect(resultDoc).toEqualDocument(doc);

    // check serializing
    expect(await serialize(doc)).toMatchSnapshot();
    const parsedDoc = await parse(await serialize(doc));
    expect(parsedDoc).toEqualDocument(doc);
  });

  test('table8 no row', async () => {
    const doc = (
      <doc>
        <heading>Table 8</heading>
        <table>
          <table_row>
            <table_header>
              <para>abc</para>
            </table_header>
            <table_header>
              <para>def</para>
            </table_header>
          </table_row>
        </table>
      </doc>
    );

    const md = await fs.readFile(
      path.join(__dirname, './fixtures/table8.md'),
      'utf-8',
    );
    const resultDoc = await parse(md);
    expect(resultDoc).toEqualDocument(doc);

    // check serializing
    expect(await serialize(doc)).toMatchSnapshot();
    const parsedDoc = await parse(await serialize(doc));
    expect(parsedDoc).toEqualDocument(doc);
  });
});
