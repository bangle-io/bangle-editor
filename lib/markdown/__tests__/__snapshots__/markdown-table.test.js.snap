// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown table parsing table1.md: a simple table 1`] = `
"# My Table

| NOTE:            |
|:-----------------|
| Work in progress |
"
`;

exports[`Markdown table parsing table2.md: two simple table with two columns 1`] = `
"# Table 2

| Header 1 | Header 2 |
|----------|----------|
| hello    | world    |

| Header 1 | Header 2 |
|----------|----------|
| hello    | world    |
"
`;

exports[`Markdown table parsing table3.md: with alignment 1`] = `
"# Table 3

| Default | Left | Center | Right |
|---------|:-----|:------:|------:|
| x       | x    | x      | x     |
"
`;

exports[`Markdown table parsing table4.md 1`] = `
"# Table 4

| foo | goo |
|:---:|----:|
| bar | baz |
"
`;

exports[`Markdown table parsing table5.md 1`] = `
"# Table 5

| f\\|oo    |
|---------|
| b \`\\|\` az |
| b **\\|** im  |
"
`;

exports[`Markdown table parsing table6.md 1`] = `
"# Table 6

| abc | def |
|-----|-----|
| bar | baz |
> bar"
`;

exports[`Markdown table parsing table7 missing column cell 1`] = `
"# Table 7

| abc | def | | --- | | bar |"
`;

exports[`Markdown table parsing table8 no row 1`] = `
"# Table 8

| abc | def |
|-----|-----|"
`;
